src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java

```diff
@@ -26,15 +30,26 @@
 public class CopyDoneMessage extends ControlMessage {
 
   protected static final char IDENTIFIER = 'c';
+  private CopyStatement statement;
 
   public CopyDoneMessage(ConnectionHandler connection) throws Exception {
     super(connection);
+    this.statement = (CopyStatement) connection.getActiveStatement();
   }
 
   @Override
   protected void sendPayload() throws Exception {
-    throw new IllegalStateException(
-        "Spanner does not currently support the copy functionality through the proxy.");
+    // If backend error occurred during copy-in mode, drop any subsequent CopyDone messages.
+    MutationWriter mw = this.statement.getMutationWriter();
+    if (!statement.hasException()) {
+      int rowCount = mw.writeToSpanner(this.connection); // Write any remaining mutations to Spanner
+      statement.addUpdateCount(rowCount); // Increase the row count of number of rows copied.
+      this.sendSpannerResult(this.statement, QueryMode.SIMPLE, 0L);
+      new ReadyResponse(this.outputStream, ReadyResponse.Status.IDLE).send();
+    } else {
+      mw.closeErrorFile();
```

Comment:
Shouldn't we send a `ReadyResponse` here as well? If I understand the implementation correctly, we abort the copy operation if an error occurs, and do not expect any more Copy messages, which means that we should be in the `Idle` state.

Reply:
Yes, we should send a ReadyResponse in this case.